File: day16/typescript/day16.ts
import { readFile } from 'fs';
import { promisify } from 'util';

import * as R from 'ramda';

interface Spin {
    kind: 'spin';
    data: number;
}

interface Exchange {
    kind: 'exchange';
    data: number[];
}

interface Partner {
    kind: 'partner';
    data: string[];
}

type Move = Spin | Exchange | Partner;
type Programs = string[];

const programs = Array.from('abcdefghijklmnop');

const parse = R.map<string, Move>(
    (m: string) => {
        switch (R.head(m)) {
            case 's':
                return { kind: 'spin', data: parseInt(R.tail(m), 10) };
            case 'x':
                return { kind: 'exchange', data: R.map(parseInt, R.split('/', R.tail(m))) };
            default:
                return { kind: 'partner', data: R.split('/', R.tail(m)) };
        }
    },
);

const spin = (ps: Programs, s: Spin) => R.compose<Programs, Programs[], Programs[], Programs>(
    x => R.flatten<string>(x),
    R.reverse,
    R.splitAt(ps.length - s.data),
)(ps);

const exchange = (ps: Programs, e: Exchange): Programs => {
    const [a, b] = R.map(R.lensIndex, e.data);
    return R.compose<Programs, Programs, Programs>(
        R.set<string>(b, R.view(a, ps)),
        R.set<string>(a, R.view(b, ps)),
    )(ps);
};

const partner = (ps: Programs, p: Partner): Programs => {
    return exchange(ps, {
        kind: 'exchange',
        data: R.map(c => R.findIndex(R.equals(c), ps), p.data),
    });
};

const makeMoves = (ps: Programs) => R.reduce<Move, Programs>(
    (acc, curr) => {
        switch (curr.kind) {
            case 'spin':
                return spin(acc, curr);
            case 'exchange':
                return exchange(acc, curr);
            default:
                return partner(acc, curr);
        }
    },
    ps,
);

const part1 = R.compose(R.join(''), makeMoves(programs));

const part2 = (moves: Move[]) => {
    let ps = [...programs];
    let count = 0;
    do {
        ps = makeMoves(ps)(moves);
        count++;
    } while (ps.join('') !== programs.join(''));

    const rem = 1000000000 % count;
    for (const _ of R.range(0, rem)) {
        ps = makeMoves(ps)(moves);
    }

    return ps.join('');
};

(async () => {
    const input = await promisify(readFile)('day16/input.txt', 'utf8');
    const moves = parse(R.split(',', input));

    console.log(part1(moves));
    console.log(part2(moves));
})();

File: day16/haskell/day16.hs
import Control.Arrow ((&&&))
import Data.Array.Unboxed (
    UArray
    , ixmap
    , bounds
    , elems
    , listArray
    , (//)
    , (!)
    )
import Data.List (elemIndex)
import Data.List.Split (splitOn)
import Data.Maybe (fromJust)

type Programs = UArray Int Char;
data Move = Spin Int 
          | Exchange Int Int 
          | Partner Char Char 
          deriving (Show, Eq)

readMove :: String -> Move
readMove ('s':xs) = Spin (read xs)
readMove ('x':xs) = f . map read . splitOn "/" $ xs
    where f [a, b] = Exchange a b
readMove ('p':xs) = f . splitOn "/" $ xs
    where f [a:_, b:_] = Partner a b

parse :: String -> [Move]
parse = map readMove . splitOn ","

move :: Programs -> Move -> Programs
move ps (Spin n)       = ixmap (bounds ps) spin ps
    where spin i = abs $ (i - n) `mod` length (elems ps)
move ps (Exchange a b) = ps // [(a, ps ! b), (b, ps ! a)]
move ps (Partner a b)  = move ps $ Exchange (idx a) (idx b)
    where idx e = fromJust . elemIndex e . elems $ ps

makeMoves :: Programs -> [Move] -> Programs
makeMoves = foldl move

programs :: Programs
programs = listArray (0, 15) "abcdefghijklmnop"

part1 :: [Move] -> String
part1 = elems . makeMoves programs

findCycle :: [Move] -> [Programs]
findCycle ms = go [makeMoves programs ms]
    where go ps@(s:_)
            | s == programs = ps
            | otherwise     = go (makeMoves s ms : ps)

part2 :: [Move] -> String
part2 ms = elems $ c !! (l - 1000000000 `mod` l)
    where (l, c) = (length &&& id) (findCycle ms)

main :: IO ()
main = do
    input <- parse <$> readFile "day16/input.txt"
    print . part1 $ input
    print . part2 $ input